### ISSUE TITLE
How to directly access LPT in DOSBox-X under Windows 7

### ISSUE THREAD

**Issue by dtbalazs** [joncampbell123/dosbox-x#1948]
Hi,
I have just installed DOSBox-X and tried to access a PROM programmer device from its DOS driver through LPT1.
The error message is as follows: "Cannot directly access the printer ports."

My LPT config is as follows
parallel1 = reallpt realbase:378
parallel2 = disabled
parallel3 = disabled
dongle    = false

What am I doing wrong?

Thanks:
Balazs Toth

**Comment by joncampbell123** [comment]
If you are doing this on Linux, you'll need to run DOSBox-X using sudo or as root, and specify the parallel port as a device (i.e. /dev/lp0).

If you are doing this on Windows, you'll need inpout32.dll and possibly run as Administrator since you are asking DOSBox-X to touch I/O ports.

**Comment by rderooy** [comment]
I know this is on Windows, but just wanted to point out that you don't need to run as root on Linux. On most Linux distributions you just need to add your user to the ``lp`` group.

```
usermod -a -G lp myuser
```

**Comment by rderooy** [comment]
Also, this ``reallpt`` option only works with legacy parallel ports, not usb parallel adapters.

Also, depending on the system the LPT1 port can be either set to 3BC or 378. In Windows, check device manager.

**Comment by dtbalazs** [comment]
Hi Jonathan and Robert,
Thank you for your help, but I got an error having installed INPOUT32.DLL, no matter DOSBox-X was executed as administrator or not.

I have Window 7 Ultimate 32 bit version.
I downloaded INPOUT32.DLL from here:  https://wikidll.com/download/10568.
I copied the unzipped DLL beside 'dosbox-x.exe' in its default install directory C:\DOSBox-X.
I have AVG Internet Security installed on the PC. (Can this be a problem?)

The detailed error message is as follows: (language is Hugarian)
Probléma-aláírás:
  Problémaesemény neve:	APPCRASH
  Alkalmazásnév:	dosbox-x.exe
  Alkalmazásverzió:	0.0.0.0 (app. version)
  Alkalmazás időbélyegzője:	5f75681e (app. timestamp)
  Hiba – modul neve:	inpout32.dll
  Hiba – modul verziója:	0.0.0.0 (error module version)
  Hiba – modul időbélyegzője:	2a425e19 (error module timestamp)
  Kivételkód:	c0000096 (exception code)
  Kivétel – eltolás:	00005c1e (exception offset)
  Operációs rendszer verziója:	6.1.7601.2.1.0.256.1 (OS version)
  Területibeállítás-azonosító:	1038 (language)
  További információk 1:	0a9e
  További információk 2:	0a9e372d3b4ad19135b953a78882e789
  További információk 3:	0a9e
  További információk 4:	0a9e372d3b4ad19135b953a78882e789

Could help me to solve this problem?
Thanks:
Balazs Toth

**Comment by rderooy** [comment]
I have no experience with inpout32.dll so cannot help. Also I have no parallel port/cable/devices any longer, so I have no way of testing this.

But you did not answer my questions. Is this a real legacy parallel port, or a USB to Parallel adapter? And if it is a legacy parallel port, what IO port does Windows Device Manager say it is set to?

**Comment by cracyc** [comment]
Might be neat to add libusb/winusb support for usb parallel adapters so printers and other parallel devices can used without needing any kernel driver or root access.

**Comment by Wengier** [comment]
@cracyc Yes, that is nice to have of course. Alternatively, there is the LPT2USB device (https://www.lpt2usb.net/) which is designed to interface the PC parallel port to USB printer port, thereby enabling USB printers to work with the LPT ports. I have tried this myself on one of my old computers and it works.

**Comment by Wengier** [comment]
@dtbalazs I have checked the inpout32.dll file you mentioned. It seems that the file you downloaded from there has issues and will crash. You can check out the following version instead:

http://www.highrez.co.uk/scripts/download.asp?package=InpOutBinaries

It contains both 32-bit and 64-bit binaries. Be sure the extracted dll file is named inpout32.dll. Please check it out and see if it works for you. Thanks!

**Comment by Wengier** [comment]
@dtbalazs Does the new INPOUT32.DLL file work for you now? I certainly hope so.

**Comment by dtbalazs** [comment]
Hi,

Thanks to all especially to Wengier helping me so much in this problem.
In short the EPROM programmer is working now.

Nevertheless I had some things to learn to.
Possibly I missed to read the relevant documentation.
1.
The crashing version of INPOUT.DLL.
I found the  same version what Wengier proposed to use, but this version still did not work because of the reasons listed in 2.
2.
Some steps, which I had to perform during installation, were not performed.
a. First "InstallDrive.exe" of the downloaded ZIP must be executed, which installs "inpout32.sys" in directory "c:\Windows\System32\drivers\". - This is what I missed to perform.
b. Next "inpout32.dll" of the downloaded ZIP must be copied beside "dosbox-x.exe" in its install directory.

One more question in the end.
The stuff is installed on a PC under Widows 7 Ultimate 32 bit, but I am going to install it on another PC under Windows 10 Pro 64 bit.
Though INPOUT.DLL has a 64 bit version as well, I think I still have to install its 32 bit version, because the application (dosbox-x.exe), which is going to use it, is a Win32 application.

Thanks a lot again.
Balazs Toth

**Comment by Wengier** [comment]
@dtbalazs Great to hear that it works for you.

About 32-bit vs 64-bit DLL, it really depends on the DOSBox-X build you are using, and **not** the Windows build you are using. If you are using 32-bit DOSBox-X binary, then the 32-bit DLL should be used, and if you are using 64-bit DOSBox-X binary, then the 64-bit DLL should be used. For 32-bit Windows, only 32-bit DOSBox-X binary can be run, but for 64-bit Windows, both 32-bit and 64-bit DOSBox-X binaries can be run. So use the DLL according to your DOSBox-X build.

Meanwhile, I have improved the Windows installer to include the inpout32.dll and inpoutx64.dll files so that they will be supplied to DOSBox-X by default. 64-bit DOSBox-X binaries will try to load the inpoutx64.dll file in addition to the inpout32.dll file.

**Comment by dtbalazs** [comment]
Hi Wengier,

Thanks for the answer. I think I shall use the Win32 version under both OSs

It is great that DOSBox-X will contain complete Parallel port  support.
Just to make sure. Will it contain the installation of "inpout32.sys" as well?

It was great help to make me able to directly access the parallel port.

Thanks again.
Balazs

**Comment by Wengier** [comment]
@dtbalazs Yes, the updated Windows installer will contain the inpout32.sys (and also inpoutx64.dll for 64-bit systems) file in a “typical” or “full” installation. This will allow users to use the direct parallel port feature in DOSBox-X without having to find the DLL themselves.

The best thing is that feature is now confirmed to work for you, which is excellent indeed.

**Comment by zomgugoff** [comment]
An additional note about the realbase setting; If you install a PCI or PCIe parallel port card, you will need to check the Device Manager in Windows to get the I/O port it uses and put that in the .ini file. It will not be one of the default ports.

**Comment by pppq** [comment]
Just a heads up, I understand that there is nothing really to be done on your end: some AV scanners trigger because `inpout` requires/provides low-level access (detected as "vulndriver" - the name is a bit misleading in that this could be as simple as the driver is not signed or its signature is out of date).

Example: https://www.virustotal.com/gui/file/5f27ed4d5cd58a1ee23deeb802e09e73f3a1d884ce2135f6e827f67b171269e7/detection
